### ISSUE TITLE
Async handler: max sleep_for_millis dramatically affects performance

### ISSUE THREAD

**Issue by rayrapetyan** [gabime/spdlog#698]
I've noticed something affecting performance of app working on a moderate logging rate and using async logger:

file include/spdlog/details/async_log_helper.h, function sleep_or_yeld, when max value there is defined as 500 (return details::os::sleep_for_millis(500);), it produces huge delays in app output\execution:

[2018-04-28 17:43:08.610] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:43:09.337] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:43:09.337] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:43:10.064] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:43:10.789] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:43:10.789] [root] [debug] [CheckSource:239] blablabla

With value defined as 200 (return details::os::sleep_for_millis(200);)  it works without significant delays:

[2018-04-28 17:36:01.480] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:36:01.480] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:36:01.501] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:36:01.501] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:36:01.522] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:36:01.522] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:36:01.544] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:36:01.544] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:36:01.566] [root] [debug] [CheckSource:239] blablabla
[2018-04-28 17:36:01.566] [root] [debug] [CheckSource:239] blablabla

Yes, just this single number makes this difference, and 500 is an unlucky delay value, weird a bit...

It was 200 in 0.14.0 and was changed to 500 starting from 0.16.0. I'm getting a lot of timeouts in the internal app logic because of that.

**Comment by gabime** [comment]
Right. I need to fix that somehow. Probably reduce to 200 or 100 when trying to push to the q (as oposed when trying to pop).

Why are you using such small queue  ? Try increasing it to a bigger number (128 or above)

**Comment by rayrapetyan** [comment]
2 is for a DEBUG env. I believe at some moment in past there was a logic in spdlog of not printing anything until queue is full, so often when app crashed it didn't print anything, with queue length 2 I reduced the max amount of non-printed messages.

**Comment by gabime** [comment]
> until queue is full

I am not aware of such behaviour in the past.
Just use a bigger q size or use sync logger if you cant afford losing messages on crash.

**Comment by rayrapetyan** [comment]
Maybe you’re right, the issue was with g3log, we’ve copy/pasted logic from.
But wondering why there is such a huge difference in a described case?

**Comment by gabime** [comment]
Probably because 500 ms simply makes the producer/consumer wait too much - such small queue size means it is always full or empty.
